src/nnet3/discriminative-supervision.h

```diff
@@ -0,0 +1,247 @@
+struct DiscriminativeSupervision {
+  // The weight we assign to this example;
+  // this will typically be one, but we include it
+  // for the sake of generality.  
+  BaseFloat weight; 
+  
+  // num_sequences will be 1 if you create a DiscriminativeSupervision object from a single
+  // lattice or alignment, but if you combine multiple DiscriminativeSupervision objects
+  // the 'num_sequences' is the number of objects that were combined (the
+  // lattices get appended).
+  int32 num_sequences;
+
+  // the number of frames in each sequence of appended objects.  num_frames *
+  // num_sequences must equal the path length of any path in the lattices.
+  // Technically this information is redundant with the lattices, but it's convenient
+  // to have it separately.
+  int32 frames_per_sequence;
+  
+  // The numerator alignment
+  // Usually obtained by aligning the reference text with the seed neural
+  // network model; can be the best path of generated lattice in the case of
+  // semi-supervised training.
+  std::vector<int32> num_ali;
+  
+  // Note: any acoustic
+  // likelihoods in the lattices will be
+  // recomputed at the time we train.
+  
+  // The denominator lattice.  
+  Lattice den_lat; 
+  
+  DiscriminativeSupervision(): weight(1.0), num_sequences(1),
+                               frames_per_sequence(-1) { }
+
+  DiscriminativeSupervision(const DiscriminativeSupervision &other);
+
+  void Swap(DiscriminativeSupervision *other);
+
+  bool operator == (const DiscriminativeSupervision &other) const;
+  
+  // This function checks that this supervision object satifsies some
+  // of the properties we expect of it, and calls KALDI_ERR if not.
+  void Check() const;
+  
+  inline int32 NumFrames() const { 
+    return num_sequences * frames_per_sequence; 
+  }
+
+  void Write(std::ostream &os, bool binary) const;
+  void Read(std::istream &is, bool binary);
+};
+
+// This function creates a supervision object from numerator alignment
+// and denominator lattice.  The supervision object is used for sequence
+// discriminative training.
+// Topologically sorts the lattice after copying to the supervision object.
+bool LatticeToDiscriminativeSupervision(
+    const std::vector<int32> &alignment,
+    const Lattice &lat,
+    BaseFloat weight,
+    DiscriminativeSupervision *supervision);
+
+// This class is used for splitting something of type
+// DiscriminativeSupervision into
+// multiple pieces corresponding to different frame-ranges.
+class DiscriminativeSupervisionSplitter {
+ public:
+  typedef fst::ArcTpl<LatticeWeight> LatticeArc;
+  typedef fst::VectorFst<LatticeArc> Lattice;
+ 
+  DiscriminativeSupervisionSplitter(
+      const SplitDiscriminativeSupervisionOptions &config,
+      const TransitionModel &tmodel,
+      const DiscriminativeSupervision &supervision);
+
+  // A structure used to store the forward and backward scores 
+  // and state times of a lattice
+  struct LatticeInfo {
+    std::vector<double> alpha_p;
```

Comment:
ok, well change it-- in any case that would have had to be documented.

On Mon, Mar 7, 2016 at 7:11 PM, Vimal Manohar <notifications@github.com>
wrote:

> In src/nnet3/discriminative-supervision.h
> <https://github.com/kaldi-asr/kaldi/pull/563#discussion_r55296622>:
>
> > +// DiscriminativeSupervision into
> > +// multiple pieces corresponding to different frame-ranges.
> > +class DiscriminativeSupervisionSplitter {
> > + public:
> > +  typedef fst::ArcTpl<LatticeWeight> LatticeArc;
> > +  typedef fst::VectorFst<LatticeArc> Lattice;
> > +
> > +  DiscriminativeSupervisionSplitter(
> > +      const SplitDiscriminativeSupervisionOptions &config,
> > +      const TransitionModel &tmodel,
> > +      const DiscriminativeSupervision &supervision);
> > +
> > +  // A structure used to store the forward and backward scores
> > +  // and state times of a lattice
> > +  struct LatticeInfo {
> > +    std::vector<double> alpha_p;
>
> This was because we had considered experimenting if we needed to also add
> initial and final costs of the objective, which would have been alpha_r and
> beta_r.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/563/files#r55296622>.
>